File: lib/sasl.c
/* Using gsasl for authentification */


#include <gsasl.h>


static Gsasl *ctx = NULL;
static Gsasl_session *session;


static int sasl_init(CHOICE_POINT *C0) 
{
  return gsasl_init(&ctx) == GSASL_OK;
}


static int sasl_done(CHOICE_POINT *C0)
{
  gsasl_done(ctx);
  return 1;
}


PRIMITIVE(sasl_client_start, X mech, X user, X passwd, X token, X realm) 
{
  int len;
  int rc = gsasl_client_start(ctx, to_string(mech, &len), &session);

  if(rc != GSASL_OK)
    system_error((XCHAR *)gsasl_strerror(rc));

  gsasl_property_set(session, GSASL_AUTHID, to_string(user, &len));
  gsasl_property_set(session, GSASL_PASSWORD, to_string(passwd, &len));
  gsasl_property_set(session, GSASL_ANONYMOUS_TOKEN, to_string(token, &len));
  gsasl_property_set(session, GSASL_REALM, to_string(realm, &len));
  return 1;
}


static int sasl_finish(CHOICE_POINT *C0)
{
  gsasl_finish(session);
  return 1;
}


PRIMITIVE(sasl_step64, X input, X result)
{
  XCHAR *p;
  int len;
  int rc = gsasl_step64(session, to_string(input, &len), &p);

  if(rc == GSASL_NEEDS_MORE || rc == GSASL_OK) {
    // succeeds if more is needed
    rc = unify(result, CSYMBOL(p));
    gsasl_free(p);
    return 1;
  }

  system_error((XCHAR *)gsasl_strerror(rc));
  return 0;			/* never executed */
}

File: lib/sasl.pl
%%%% GSASL interface - prolog part
%
% Note:
%
% - the number of steps is user-defined, dependent on the number
%   of calls to sasl_step64/2
% - GSASL-sessions can not be run in parallel.


:- verbatim('#include "sasl.c"').


sasl_init :- foreign_call(sasl_init).
sasl_done :- foreign_call(sasl_done).

sasl_client_start(MECH, USER, PASSWD, TOKEN, REALM) :-
	foreign_call(sasl_client_start(MECH, USER, PASSWD, TOKEN, REALM)).

sasl_finish :- foreign_call(sasl_finish).

sasl_step64(IN, OUT) :-
	foreign_call(sasl_step64(IN, OUT)).
